src/OpenTelemetry/Metrics/MetricStreamIdentity.cs

```diff
@@ -0,0 +1,148 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics.Metrics;
+
+namespace OpenTelemetry.Metrics
+{
+    internal readonly struct MetricStreamIdentity : IEquatable<MetricStreamIdentity>
+    {
+        private static readonly StringArrayEqualityComparer StringArrayComparer = new StringArrayEqualityComparer();
+        private readonly int hashCode;
+
+        public MetricStreamIdentity(Meter meter, string instrumentName, string unit, string description, Type instrumentType, string[] tagKeys, double[] histogramBucketBounds)
+        {
+            this.MeterName = meter.Name;
+            this.MeterVersion = meter.Version ?? string.Empty;
+            this.InstrumentName = instrumentName;
+            this.Unit = unit ?? string.Empty;
+            this.Description = description ?? string.Empty;
+            this.InstrumentType = instrumentType;
+
+            if (tagKeys != null && tagKeys.Length > 0)
+            {
+                this.TagKeys = new string[tagKeys.Length];
+                tagKeys.CopyTo(this.TagKeys, 0);
+            }
+            else
+            {
+                this.TagKeys = null;
+            }
+
+            if (histogramBucketBounds != null && histogramBucketBounds.Length > 0)
+            {
+                this.HistogramBucketBounds = new double[histogramBucketBounds.Length];
+                histogramBucketBounds.CopyTo(this.HistogramBucketBounds, 0);
+            }
+            else
+            {
+                this.HistogramBucketBounds = null;
+            }
+
+            unchecked
+            {
+                var hash = 17;
+                hash = (hash * 31) + this.InstrumentType.GetHashCode();
+                hash = (hash * 31) + this.MeterName.GetHashCode();
+                hash = (hash * 31) + this.MeterVersion.GetHashCode();
+                hash = (hash * 31) + this.InstrumentName.GetHashCode();
+                hash = this.Unit == null ? hash : (hash * 31) + this.Unit.GetHashCode();
+                hash = this.Description == null ? hash : (hash * 31) + this.Description.GetHashCode();
+                hash = this.TagKeys == null ? hash : StringArrayComparer.GetHashCode(this.TagKeys);
+                if (this.HistogramBucketBounds != null)
+                {
+                    var len = this.HistogramBucketBounds.Length;
+                    for (var i = 0; i < len; ++i)
+                    {
+                        hash = (hash * 31) + this.HistogramBucketBounds[i].GetHashCode();
+                    }
+                }
+
+                this.hashCode = hash;
+            }
+        }
+
+        public readonly string MeterName { get; }
```

Comment:
I don't think we need `readonly` on all the members because the struct itself is `readonly`. Any harm in having it? Nothing major that I can think of. Might lead to a bigger binary, I'm assuming `readonly` ends up a hint in the IL metadata. More code to maintain 😄